### ISSUE TITLE
Consider not converting spaces to &nbsp; entities in kind=markdown

### ISSUE THREAD

**Issue by cpbotha** [Microsoft/python-language-server#422]
In getting the Microsoft PyLS working with Emacs, see https://vxlabs.com/2018/11/19/configuring-emacs-lsp-mode-and-microsofts-visual-studio-code-python-language-server/ -- I had to hack around the fact that PyLS _converts_ all spaces in the Python source docstrings to `&nbsp;` elements.

As far as I can tell, this is strictly speaking valid Markdown, although it's making life difficult for non-HTML based MarkDown renderers, like those in VIM and Emacs.

**Would you consider disabling this function, or at least making it configurable?**

In short, I would like contents with kind=markdown to use simple spaces instead of `&nbsp;` entities.

**Comment by cpbotha** [comment]
I understand that you're being constrained to a certain extent by Visual Studio Code.

However, we've all seen zillions of markdown files, most of which don't contain this many `&nbsp;` entities. Markdown renderers the world over seem to do a great job of showing these `&nbsp;`-free files. :)

I would already be very happy if you could make an exception or configuration for *only* `&nbsp;`, but I of course understand if you make the call that's not going to be reasonable for you to add.

With regard to your last question: Even when I specify `preferredFormat` as `plaintext`, I get those `&nbsp;` entities, but only in some cases. For example, with `import pandas as pd`, hovering over the `pandas` will get me `plaintext`, whilst hovering over the `pd` will get me the same docs as `markdown`. Super strange.

**Comment by jakebailey** [comment]
Oh, I didn't mean that we wouldn't want to change things; I just wanted to write down my thoughts as to why I believe this happens to be the current behavior and some places to look and challenges to consider when writing a fix.

Those `&nbsp;`s shouldn't be showing up in the plaintext output, for sure. Are you also certain that the client is sending plaintext in the client capabilities? We currently have two different places to set the documentation formatting (see #319).

**Comment by cpbotha** [comment]
I'm sending `preferredFormat: plaintext` as part of the `initializationOptions`, just like vs.code is doing (but they are sending markdown)

Doing this affects some things but not others. I can confirm that I still get markdown documentation as a response to textDocument/hover -- I looked at the raw json coming back from the LS.

**Comment by MikhailArkhipov** [comment]
Plaintext means no colors in VS Code. Since extension with Jedi traditionally provided colored code (as do other extensions), we should be doing the same.

**Comment by cpbotha** [comment]
@MikhailArkhipov The point was that even when the LS is instructed via its API to return plaintext, it still returns markdown and `&nbsp;` entities, sometimes.

Also, I don't understand your reasoning with regards to TABs and `&nbsp` vs just spaces. In both cases you have to make the call how many spaces to use. `&nbsp;`-entities only serve to act as further LS-imposed constraints on the vs.code markdown renderer (and other more compliant markdown renderers). (or are you translating TABs into another more specific html entity?)

Furthermore, from a software engineering standpoint, this really goes against separation of concerns. Ideally the MS LS should generate fully compliant, textbook markdown, whilst vs.code and other clients simply should simply render that compliant, textbook markdown.

Although markdown strictly speaking might allow for `&nbsp;` and other html entities, that was never its intention. Remember that it started as a plaintext markup that could be equally well parsed by humans reading emails.

As I mentioned earlier: Fundamentally, most (all?) markdown renderers render `&nbsp;`-less markdown without a hitch. I find it hard to believe that vs.code is the only one that has trouble with this.

All of that being said: I don't really care how you decide to deviate from what everybody else is doing, as long as you stay consistent in how you do it so I don't have to keep on updating my patches to Emacs. ;P

**Comment by MikhailArkhipov** [comment]
LS sends `&nbsp;` in markdown triple-tick blocks with `html` language specified so they are part of HTML fragment. Works same way triples work here at GH. Typically markdown looks like block of triple-tick-python followed by triple-tick-html. See https://github.com/Microsoft/vscode/issues/49018 for an example.

**Comment by jakebailey** [comment]
VS Code always requests markdown, never plaintext, so I don't think fixing plaintext output to actually be plaintext will affect VS Code (not that plaintext can't contain `&nbsp;` if it means it). I'm not sure how it would affect VS, though, if at all.

I'm a bit confused what triple backtick `html` blocks have to do with this, though. That's for code formatting HTML (like triple backtick then `python` would be). Example:

```html
<body>
<p>Hello &nbsp; world!
</body>
```

Is:

````
```html
<body>
<p>Hello &nbsp; world!
</body>
```
````

That doesn't mean that HTML is rendered, but that it's highlighted as HTML. When I was looking into #328, I didn't see us outputting triple backtick `html` blocks at any point. Searching the repo for "html" doesn't net anything either. When we render

If we want that bottom docstring to be spaced as plaintext with the original characters, I'd think that a plain triple backtick block would work, or explicitly stating `plaintext`. IIRC we do the former.

**Comment by jakebailey** [comment]
All backtick blocks should be able to do that, like:

```
        I'm prefaced by 8 spaces!
```

Which is:

````
```
        I'm prefaced by 8 spaces!
```
````

**Comment by jakebailey** [comment]
With #596, the docstring handling has been completely rewritten for the new language server. There is no longer any manual spacing of elements, instead converting from the docstring to something markdown will be able to format. I don't think we'll be inserting any more HTML-specific whitespace (unless it can't be avoided, which I can't see happening). If plaintext is specified, all docstrings returned will come through unmodified, other than removing consistent indents.

I'll file another bug for handling LSP capability-based MarkupKind support, but that's a further improvement.
